insonmnia/hardware/hardware.go

```diff
@@ -146,56 +147,141 @@ func (m *benchValue) Value() uint64 {
+// TODO: find a way to refactor all this shit.
+func (h *Hardware) LimitTo(resources *sonm.AskPlanResources) (*Hardware, error) {
+	hardware := &Hardware{
+		CPU: &sonm.CPU{Device: h.CPU.Device, Benchmarks: map[uint64]*sonm.Benchmark{}},
+		GPU: []*sonm.GPU{},
+		RAM: &sonm.RAM{Device: h.RAM.Device, Benchmarks: map[uint64]*sonm.Benchmark{}},
+		Network: &sonm.Network{
+			In:       h.Network.In,
+			Out:      h.Network.Out,
+			Overlay:  h.Network.Overlay,
+			Incoming: h.Network.Incoming,
+			Outbound: h.Network.Outbound,
+		},
+		Storage: &sonm.Storage{Device: h.Storage.Device, Benchmarks: map[uint64]*sonm.Benchmark{}},
+	}
+
+	for _, gpu := range h.GPU {
+		hardware.GPU = append(hardware.GPU, &sonm.GPU{Device: gpu.Device, Benchmarks: map[uint64]*sonm.Benchmark{}})
+	}
+
+	for _, hash := range resources.GetGPU().GetHashes() {
+		found := false
+		for idx, gpu := range h.GPU {
+			if gpu.GetDevice().GetHash() == hash {
+				for _, bench := range gpu.GetBenchmarks() {
+					insertBench(hardware.GPU[idx].Benchmarks, bench, 1.0)
+				}
+				found = true
+				break
 			}
 		}
+		if !found {
+			return nil, fmt.Errorf("unknown hash in passed resources: %s", hash)
+		}
+	}
+
+	proportion := float64(resources.GetCPU().GetCorePercents()) / float64(h.CPU.GetDevice().GetCores()) / 100
+	for _, bench := range h.CPU.GetBenchmarks() {
+		insertBench(hardware.CPU.Benchmarks, bench, proportion)
+	}
 
+	proportion = float64(resources.GetStorage().GetSize().GetBytes()) / float64(h.Storage.GetDevice().GetBytesAvailable())
+	for _, bench := range h.Storage.GetBenchmarks() {
+		insertBench(hardware.Storage.Benchmarks, bench, proportion)
 	}
-	return to, nil
+
+	proportion = float64(resources.GetRAM().GetSize().GetBytes()) / float64(h.RAM.GetDevice().GetAvailable())
+	for _, bench := range h.RAM.GetBenchmarks() {
+		insertBench(hardware.RAM.Benchmarks, bench, proportion)
+	}
+
+	proportion = float64(resources.GetNetwork().GetThroughputIn().GetBitsPerSecond()) / float64(h.Network.GetIn())
+	for _, bench := range h.Network.GetBenchmarksIn() {
+		insertBench(hardware.Network.BenchmarksIn, bench, proportion)
+	}
+
+	proportion = float64(resources.GetNetwork().GetThroughputOut().GetBitsPerSecond()) / float64(h.Network.GetOut())
+	for _, bench := range h.Network.GetBenchmarksOut() {
+		insertBench(hardware.Network.BenchmarksOut, bench, proportion)
+	}
+
+	return hardware, nil
 }
 
-func (h *Hardware) ResourcesToBenchmarks(resources *sonm.AskPlanResources) (*sonm.Benchmarks, error) {
+func (h *Hardware) ResourcesToBenchmarkMap(resources *sonm.AskPlanResources) (benchmarks map[uint64]*sonm.Benchmark, err error) {
 	if !resources.GPU.Normalized() {
 		return nil, errors.New("passed resources are not normalized, call resources.GPU.Normalize(hardware) first")
 	}
-	var err error
-	benchmarks := make([]benchValue, sonm.MinNumBenchmarks)
 
-	proportions := []float64{}
-	hwBenches := []map[uint64]*sonm.Benchmark{}
+	defer func() {
```

Comment:
is this exception handling that we deserved?

Reply:
yup